packages/pkl.tar/File.pkl

```diff
@@ -0,0 +1,407 @@
+local const ustarMaxSizeTimestamp: UInt = 2 ** 33 as UInt
+
+/// File type marker, also known as `linkflag`.
+fixed type: Char
+
+/// The numeric byte value of [type].
+fixed typeByte: UInt8 = type.encodeToBytes("UTF-8")[0]
+
+/// File information.
+info: Info?
+
+/// Pax archive extended header data fields.
+paxFields: PaxFields = new {
+  when (info != null) {
+    when (info.gid > ustarMaxUidGid) {
+      ["gid"] = info.gid.toString()
+    }
+    when (info.uid > ustarMaxUidGid) {
+      ["uid"] = info.uid.toString()
+    }
+    when (info.mtime != info.mtime.truncate()) {
+      ["mtime"] = formatPaxNumber(info.mtime)
+    }
+    when (!(info.uname is PosixPortableString)) {
+      ["uname"] = info.uname
+    }
+    when (!(info.gname is PosixPortableString)) {
+      ["gname"] = info.gname
+    }
+  }
+}
+
+/// The payload bytes of this file, padded to the nearest 512 bytes.
+fixed paddedBytes: Buffer(length % blockLength == 0) = List()
+
+/// Generate the UStar header for this file including [type]-specific customizations.
+function makeHeader(path: String, pax: tar.Pax?): Block = makeBaseHeader(path, pax)
+
+/// Generate the UStar header for this file before [type]-specific customatizations.
+function makeBaseHeader(path: String, pax: tar.Pax?): Block =
+  let (pathParts = splitFilename(path, pax))
+    List(
+      pathParts.first, // name [0, 100)
+      formatOctal(info!!.encodedMode, 8), // mode [100, 108)
+      formatOctal(validateNumericField(info!!.uid, ustarMaxUidGid, "uid", pax), 8), // uid [108, 116)
+      formatOctal(validateNumericField(info!!.gid, ustarMaxUidGid, "gid", pax), 8), // gid [116, 124)
+      nul(12), // size [124, 136)
+      formatOctal(
+        validateNumericField(info!!.mtime.toInt(), ustarMaxSizeTimestamp, "mtime", pax),
+        12
+      ), // mtime [136, 148)
+      nul(8), // cksum [148, 156)
+      List(typeByte), // typeflag [156, 157)
+      nul(100), // linkname [157, 257)
+      magic, // magic [257, 263)
+      version, // version [263, 265)
+      validateStringField(info!!.uname, 32, "uname", pax), // uname [265, 297)
+      validateStringField(info!!.gname, 32, "gname", pax), // gname [297, 329)
+      nul(8), // devmajor [329, 337)
+      nul(8), // devminor [337, 345)
+      pathParts.second, // prefix [345, 500)
+      nul(12) // padding [500, 512)
+    ).flatten()
+
+local self = this
+
+/// Generate the pax header and payload for this file, if enabled and applicable.
+function makePax(path: String, pax: tar.Pax): Buffer(length % blockLength == 0) =
+  let (
+    paxHeader = new ExtendedHeader {
+      paxFields = (self.paxFields) {
+        when (splitFilename(path, pax) == null) {
+          ["path"] = path
+        }
+      }
+    }
+  )
+    if (paxHeader.paxFields.isEmpty)
+      List()
+    else
+      checksum(paxHeader.makeHeader(path, pax)) + paxHeader.paddedBytes
+
+/// File metdata information.
+class Info {
+  /// File permissions mode.
+  mode: Set<Mode>
```

Comment:
You can also use octal literals in Pkl, so, you could just leave this as type `Int(isBetween(0o000, 0o777))`; and users would just do `0o444` etc.